catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/DescribableServiceMap.java

```diff
@@ -15,29 +15,46 @@
 
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Dictionary;
 import java.util.HashMap;
+import java.util.Hashtable;
 import java.util.Map;
 import java.util.Set;
 
 import org.osgi.framework.Bundle;
 import org.osgi.framework.BundleContext;
+import org.osgi.framework.Constants;
 import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceEvent;
 import org.osgi.framework.ServiceReference;
+import org.osgi.service.event.Event;
+import org.osgi.service.event.EventConstants;
+import org.osgi.service.event.EventHandler;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import ddf.catalog.util.Describable;
 
-public class DescribableServiceMap<V extends Describable> implements Map<String, V> {
+public class DescribableServiceMap<V extends Describable> implements Map<String, V>, EventHandler {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(DescribableServiceMap.class);
 
     private static final String READ_ONLY_ERROR_MESSAGE = "This map is meant to be read only.";
 
     private Map<String, V> serviceMap = Collections.synchronizedMap(new HashMap<>());
 
+    private Map<String, String> pidToNameMap = Collections.synchronizedMap(new HashMap<>());
+
+    public DescribableServiceMap() {
+        String[] topics = new String[] {"org/osgi/framework/ServiceEvent/MODIFIED"};
+
+        Dictionary props = new Hashtable();
+        props.put(EventConstants.EVENT_TOPIC, topics);
+        getContext().registerService(EventHandler.class.getName(), this, props);
```

Comment:
Registering is normally done in blueprint but I have put it here to make the DescribableServiceMap easier to use and to avoid any misconfiguration.